### DISCUSSION TITLE
Parallel test case execution

### DISCUSSION THREAD

**Discussion by haresh-codi** [playframework/playframework#12026] [General]
Hey how to run test cases parallely in scala using play framework
Also I have added both these in `build.sbt` but still not running test cases in parallel

```sbt
fork in Test := true
parallelExecution in Test := true
```

**Comment by mkurz** [comment]
AFAIK `parallelExecution in Test := true` alone does not help for all scenarios. Take a look here:
https://github.com/playframework/playframework/blob/6ddc99c45261d1beb026fbb56e35782bd1d5ba4d/testkit/play-test/src/main/scala/play/api/test/Helpers.scala#L50-L67
When "running" a Play app or server in tests, these by default still get executed sequentially, because of `runSynchronized`. You can override `shouldRunSequentially` to change that, like we do here:
https://github.com/playframework/playframework/blob/6ddc99c45261d1beb026fbb56e35782bd1d5ba4d/core/play-integration-test/src/test/scala/play/it/http/websocket/WebSocketSpec.scala#L446
and here:
https://github.com/playframework/playframework/blob/6ddc99c45261d1beb026fbb56e35782bd1d5ba4d/core/play-integration-test/src/test/scala/play/it/http/JavaResultsHandlingSpec.scala#L44

This behaviour was added because of what's written in the comment:
> `* Tests using servers share a test server port so we default to true.`

However with Play 2.9 / 3.0 tests that open a port will run on a random port now, so we should be able to set `shouldRunSequentially` to false by default or even completely remove the `runSynchronized` method.
Not just because of that, but also now that since a couple of Play versions we do not have a global app enabled by default anymore:
https://github.com/playframework/playframework/blob/6ddc99c45261d1beb026fbb56e35782bd1d5ba4d/testkit/play-test/src/main/scala/play/api/test/TestServerFactory.scala#L56-L62
So this `runSynchronized` method comes from a time when we still used a fixed port for tests and a global app...

However AFAIK, when I worked on the random port thing a couple of months ago, and if I recall correctly, there is more stuff that needs to be done so we can switch to `false` by default (because I tried that back then :wink:), like
- #10939, which otherwise makes a lot of tests failing.

This is something to be looked into Play 2.10 and 3.1.

BTW, if you use Play Java the same is true like described above, see:
https://github.com/playframework/playframework/blob/6ddc99c45261d1beb026fbb56e35782bd1d5ba4d/testkit/play-test/src/main/java/play/test/Helpers.java#L561-L563

**Reply by mkurz** [reply]
Added an issue:
- #12214

**Comment by janani-reddy9** [comment]
Hi @mkurz, I was looking into the issues and PR's mentioned wrt this.
If I understood correctly, To have parallel running of test cases, PR - https://github.com/playframework/playframework/pull/10939 need to be worked on first.

I have a quick doubt wrt this. I saw the implementations of DevServerStart and ProdServerStart. I couldn't understand the need to initialise LoggerConfigurator before the start of application in DevServerStart?

**Comment by mkurz** [comment]
- Fixed by #13096
